File: server/golang/lib-go/entity/msg/msg.go
package msg

import (
	"encoding/binary"
	"time"
)

type Type = uint16

const (
	HeadLen            int    = 32
	ExtensionThreshold int    = 1<<6 - 1
	PayloadThreshold   int    = 1<<14 - 1
	GroupIdThreshold   uint64 = 1 << 36
)

const (
	BitMaskLeft46  uint64 = 0xFFFF_C000_0000_0000
	BitMaskRight46 uint64 = 0x00003FFFFFFFFF
	BitMaskLeft50  uint64 = 0xFFFC000000000000
	BitMaskRight50 uint64 = 0x0003FFFFFFFFFFFF
	BitMaskLeft12  uint64 = 0xFFF0000000000000
	BitMaskRight12 uint64 = 0x000FFFFFFFFFFFFF
)

const (
	NA              Type = 0
	Ack             Type = 1
	Text            Type = 32
	Meme            Type = 33
	File            Type = 34
	Image           Type = 35
	Video           Type = 36
	Audio           Type = 37
	Edit            Type = 64
	Withdraw        Type = 65
	Auth            Type = 96
	Ping            Type = 97
	Pong            Type = 98
	Echo            Type = 99
	Error           Type = 100
	BeOffline       Type = 101
	InternalError   Type = 102
	SystemMessage   Type = 128
	AddFriend       Type = 129
	RemoveFriend    Type = 130
	JoinGroup       Type = 131
	LeaveGroup      Type = 132
	RemoteInvoke    Type = 133
	SetRelationship Type = 134
	Noop            Type = 160
	Close           Type = 161
)

type Msg struct {
	inner []byte
}

type Head struct {
	versionSender          uint64
	nodeIdReceiver         uint64
	typeExtensionTimestamp uint64
	payloadSeqnum          uint64
}

type innerHead struct {
	version         uint32
	sender          uint64
	nodeId          uint32
	receiver        uint64
	typ             Type
	extensionLength uint8
	timestamp       uint64
	payloadLength   uint16
	seqnum          uint64
}

func (h *Head) toInnerHead() innerHead {
	version := uint32(h.versionSender >> 46)
	sender := h.versionSender & BitMaskRight46
	nodeId := uint32(h.nodeIdReceiver >> 46)
	receiver := h.nodeIdReceiver & BitMaskRight46
	typ := Type(h.typeExtensionTimestamp >> 52)
	extensionLength := uint8((h.typeExtensionTimestamp & BitMaskRight12) >> 46)
	timestamp := h.typeExtensionTimestamp & BitMaskRight46
	payloadLength := uint16(h.payloadSeqnum >> 50)
	seqnum := h.payloadSeqnum & BitMaskRight50
	return innerHead{
		version:         version,
		sender:          sender,
		nodeId:          nodeId,
		receiver:        receiver,
		typ:             typ,
		extensionLength: extensionLength,
		timestamp:       timestamp,
		payloadLength:   payloadLength,
		seqnum:          seqnum,
	}
}

func (i *innerHead) toHead() Head {
	versionSender := uint64(i.version) << 46
	versionSender |= i.sender
	nodeIdReceiver := uint64(i.nodeId) << 46
	nodeIdReceiver |= i.receiver
	typeExtensionTimestamp := uint64(i.typ) << 52
	typeExtensionTimestamp |= uint64(i.extensionLength) << 46
	typeExtensionTimestamp |= i.timestamp
	payloadSeqnum := uint64(i.payloadLength) << 50
	payloadSeqnum |= i.seqnum
	return Head{
		versionSender:          versionSender,
		nodeIdReceiver:         nodeIdReceiver,
		typeExtensionTimestamp: typeExtensionTimestamp,
		payloadSeqnum:          payloadSeqnum,
	}
}

func (i *innerHead) toBytes() []byte {
	versionSender := uint64(i.version) << 46
	versionSender |= i.sender
	nodeIdReceiver := uint64(i.nodeId) << 46
	nodeIdReceiver |= i.receiver
	typeExtensionTimestamp := uint64(i.typ) << 52
	typeExtensionTimestamp |= uint64(i.extensionLength) << 46
	typeExtensionTimestamp |= i.timestamp
	payloadSeqnum := uint64(i.payloadLength) << 50
	payloadSeqnum |= i.seqnum
	bytes := make([]byte, HeadLen, HeadLen)
	binary.BigEndian.PutUint64(bytes[0:8], versionSender)
	binary.BigEndian.PutUint64(bytes[8:16], nodeIdReceiver)
	binary.BigEndian.PutUint64(bytes[16:24], typeExtensionTimestamp)
	binary.BigEndian.PutUint64(bytes[24:32], payloadSeqnum)
	return bytes
}

func Bytes2Head(bytes []byte) Head {
	versionSender := binary.BigEndian.Uint64(bytes[0:8])
	nodeIdReceiver := binary.BigEndian.Uint64(bytes[8:16])
	typeExtensionTimestamp := binary.BigEndian.Uint64(bytes[16:24])
	payloadSeqnum := binary.BigEndian.Uint64(bytes[24:32])
	return Head{
		versionSender:          versionSender,
		nodeIdReceiver:         nodeIdReceiver,
		typeExtensionTimestamp: typeExtensionTimestamp,
		payloadSeqnum:          payloadSeqnum,
	}
}

func PreAlloc(head *Head) Msg {
	extensionLength := int((head.typeExtensionTimestamp & BitMaskRight12) >> 46)
	payloadLength := int(head.payloadSeqnum >> 50)
	inner := make([]byte, HeadLen+extensionLength+payloadLength, HeadLen+extensionLength+payloadLength)
	binary.BigEndian.PutUint64(inner[0:8], head.versionSender)
	binary.BigEndian.PutUint64(inner[8:16], head.nodeIdReceiver)
	binary.BigEndian.PutUint64(inner[16:24], head.typeExtensionTimestamp)
	binary.BigEndian.PutUint64(inner[24:32], head.payloadSeqnum)
	return Msg{inner: inner}
}

func (m *Msg) AsBytes() []byte {
	return m.inner
}

func TextMsg(sender uint64, receiver uint64, nodeId uint32, text string) Msg {
	innerHead := innerHead{
		version:         1,
		sender:          sender,
		nodeId:          nodeId,
		receiver:        receiver,
		typ:             Text,
		extensionLength: 0,
		timestamp:       uint64(time.Now().UnixMilli()),
		payloadLength:   uint16(len(text)),
		seqnum:          0,
	}
	inner := make([]byte, HeadLen+len(text), HeadLen+len(text))
	copy(inner[0:HeadLen], innerHead.toBytes())
	copy(inner[HeadLen:], text)
	return Msg{inner: inner}
}

File: ui-prototype/src/entity/msg.ts
import { timestamp } from "../util/base";

// @ts-ignore
const BIT_MASK_LEFT_46: bigint = 0xFFFF_C000_0000_0000n;
const BIT_MASK_RIGHT_46: bigint = 0x0000_3FFF_FFFF_FFFFn;
// @ts-ignore
const BIT_MASK_LEFT_50: bigint = 0xFFFC_0000_0000_0000n;
const BIT_MASK_RIGHT_50: bigint = 0x0003_FFFF_FFFF_FFFFn;
// @ts-ignore
const BIT_MASK_LEFT_12: bigint = 0xFFF0_0000_0000_0000n;
const BIT_MASK_RIGHT_12: bigint = 0x000F_FFFF_FFFF_FFFFn;

const HEAD_LEN = 32;
// @ts-ignore
const EXTENSION_THRESHOLD = 1 << 6 - 1;
// @ts-ignore
const PAYLOAD_THRESHOLD = 1 << 14 - 1;
// user_id lager than(also equal) this value is considered as a group
const GROUP_ID_THRESHOLD: bigint = BigInt(1 << 36);

enum Type {
    NA = 0,
    /// this type can only be used for acknowledging certain msg.
    /// it's so special that we put it on the top of the enum.
    Ack = 1,

    /// the below types are used for user's communication.
    ///
    /// pure message part
    Text = 32,
    Meme = 33,
    File = 34,
    Image = 35,
    Video = 36,
    Audio = 37,
    /// control message part
    Edit = 64,
    Withdraw = 65,

    /// the below types are used for user and server's communication.
    ///
    /// logic part
    Auth = 96,
    Ping = 97,
    Pong = 98,
    Echo = 99,
    Error = 100,
    BeOffline = 101,
    InternalError = 102,
    /// business part
    /// some types may derived by user but send between server, those types are also viewed as business type.
    SystemMessage = 128,
    AddFriend = 129,
    RemoveFriend = 130,
    JoinGroup = 131,
    LeaveGroup = 132,
    RemoteInvoke = 133,
    SetRelationship = 134,
}

class Head {
    version: number;
    sender: bigint;
    nodeId: number;
    receiver: bigint;
    type: Type;
    extensionLength: number;
    timestamp: bigint;
    payloadLength: number;
    seqnum: bigint;

    constructor(
        version: number,
        sender: bigint,
        nodeId: number,
        receiver: bigint,
        type: Type,
        extensionLength: number,
        timestamp: bigint,
        payloadLength: number,
        seqNum: bigint,
    ) {
        this.version = version;
        this.sender = sender;
        this.nodeId = nodeId;
        this.receiver = receiver;
        this.type = type;
        this.extensionLength = extensionLength;
        this.timestamp = timestamp;
        this.payloadLength = payloadLength;
        this.seqnum = seqNum;
    }

    static fromArrayBuffer = (buffer: ArrayBuffer): Head => {
        let view = new DataView(buffer);
        let versionSender = view.getBigUint64(0, false);
        let nodeIdReceiver = view.getBigUint64(8, false);
        let typeExtensionLengthTimestamp = view.getBigUint64(16, false);
        let payloadLengthWithSeqNum = view.getBigUint64(24, false);
        let version = Number(versionSender >> 46n);
        let sender = BigInt(versionSender & BIT_MASK_RIGHT_46);
        let nodeId = Number(nodeIdReceiver >> 46n);
        let receiver = BigInt(nodeIdReceiver & BIT_MASK_RIGHT_46);
        let type = Number(typeExtensionLengthTimestamp >> 52n) as Type;
        let extensionLength = Number((typeExtensionLengthTimestamp & BIT_MASK_RIGHT_12) >> 46n);
        let timestamp = BigInt(typeExtensionLengthTimestamp & BIT_MASK_RIGHT_46);
        let payloadLength = Number(payloadLengthWithSeqNum >> 50n);
        let seqNum = BigInt(payloadLengthWithSeqNum & BIT_MASK_RIGHT_50);
        return new Head(
            version,
            sender,
            nodeId,
            receiver,
            type,
            extensionLength,
            timestamp,
            payloadLength,
            seqNum,
        );
    }

    toArrayBuffer(): ArrayBuffer {
        let buffer = new ArrayBuffer(HEAD_LEN);
        let view = new DataView(buffer);
        let versionSender = BigInt(this.version) << 46n | BigInt(this.sender);
        let nodeIdReceiver = BigInt(this.nodeId) << 46n | BigInt(this.receiver)
        let typeExtensionLengthTimestamp = BigInt(this.type) << 52n
            | BigInt(this.extensionLength) << 46n
            | BigInt(this.timestamp);
        let payloadLengthWithSeqNum = BigInt(this.payloadLength) << 50n | BigInt(this.seqnum);
        view.setBigUint64(0, BigInt(versionSender), false);
        view.setBigUint64(8, BigInt(nodeIdReceiver), false);
        view.setBigUint64(16, BigInt(typeExtensionLengthTimestamp), false);
        view.setBigUint64(24, BigInt(payloadLengthWithSeqNum), false);
        return buffer;
    }
}

class Msg {
    head: Head;
    payload: ArrayBuffer;
    extension: ArrayBuffer;

    constructor(head: Head, payload: ArrayBuffer, extension: ArrayBuffer) {
        this.head = head;
        this.payload = payload;
        this.extension = extension;
    }

    static fromArrayBuffer = (buffer: ArrayBuffer): Msg => {
        let head = Head.fromArrayBuffer(buffer.slice(0, HEAD_LEN));
        let payload = buffer.slice(HEAD_LEN, HEAD_LEN + head.payloadLength);
        let extension = buffer.slice(HEAD_LEN + head.payloadLength);
        return new Msg(head, payload, extension);
    }

    toArrayBuffer = (): ArrayBuffer => {
        let buffer = new ArrayBuffer(HEAD_LEN + this.head.payloadLength + this.head.extensionLength);
        let head = new Uint8Array(this.head.toArrayBuffer());
        let view = new DataView(buffer);
        for (let i = 0; i < HEAD_LEN; i++) {
            view.setUint8(i, head[i]);
        }
        let payload = new Uint8Array(this.payload);
        for (let i = 0; i < this.head.payloadLength; i++) {
            view.setUint8(HEAD_LEN + i, payload[i]);
        }
        let extension = new Uint8Array(this.extension);
        for (let i = 0; i < this.head.extensionLength; i++) {
            view.setUint8(HEAD_LEN + this.head.payloadLength + i, extension[i]);
        }
        return buffer;
    }

    payloadText = (): string => {
        return new TextDecoder().decode(this.payload);
    }

    extensionText = (): string => {
        return new TextDecoder().decode(this.extension);
    }

    static text = (sender: bigint, receiver: bigint, nodeId: number, text: string): Msg => {
        let payload = new TextEncoder().encode(text);
        let head = new Head(0, sender, nodeId, receiver, Type.Text, 0, timestamp(), payload.length, 0n);
        return new Msg(head, payload, new ArrayBuffer(0));
    }

    static text2 = (sender: bigint, receiver: bigint, nodeId: number, text: string, extension: string): Msg => {
        let payload = new TextEncoder().encode(text);
        let extensionArrayBuffer = new TextEncoder().encode(extension);
        let head = new Head(0, sender, nodeId, receiver, Type.Text, extension.length, timestamp(), payload.length, 0n);
        return new Msg(head, payload, extensionArrayBuffer);
    }

    static text0 = (sender: bigint, receiver: bigint, nodeId: number, text: string, timestamp: bigint): Msg => {
        let payload = new TextEncoder().encode(text);
        let head = new Head(0, sender, nodeId, receiver, Type.Text, 0, timestamp, payload.length, 0n);
        return new Msg(head, payload, new ArrayBuffer(0));
    }
}

export { Type, Head, Msg, GROUP_ID_THRESHOLD };